wifi_now esphome component - communicate directly to other ESP device via ESP-Now

motwok opened the pull request:
## Description:
wifi_now esphome component - communicate directly to other ESP device via ESP-Now

- send and receive supported by yaml.
- a very simple API for other components.

This PR needs #1306 to function correctly!

**Related issue (if applicable):** fixes <link to issue>
[ESP-Now integration #275](https://github.com/esphome/feature-requests/issues/275)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#775

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

motwok committed:
adding wifi_now component

motwok committed:
adding int payload
fixing linter message

motwok committed:
fix spelling error retrys
fix linter message

motwok committed:
add tests for wifi_now component

motwok committed:
fix indention in Test include

motwok committed:
fix lambda in test include

jesserockz commented:
I know this is an initial implementation, but I feel like there could be a different approach to this to make it easier to use in the ESPHome ecosystem without having to know all the details.

On each side could the sensors/binary_sensors/switches be implemented as platforms?
With one side that sets up to receive the payloads is exposed like other sensors
```
sensor:
  - platform: wifi_now
    name: "Sensor from A" 
    peer_id: peer_a
```
And the other side set up with a platform that could internally subscribe to the state of the sensor with `sensor.add_on_state_callback`
```
sensor:
  - platform: some_platform
    id: other_sensor
  - platform: wifi_now_send
    id: my_sensor
    peer_id: peer_b
    sensor: other_sensor
```

This would mean the sensors (or binary_sensor/text_sensor/etc) can be exposed directly to HA as if it belonged to `device b` which is on the WiFi. And `device a` is just an extension of sensors and switches for `device b`.

Switches could be the similar with a few minor differences as they are controlled from HA/`device b`

The retrying etc can be extracted to be behind the scenes and only surface up in an `on_error` if it really fails badly?

These are just the first things that came to mind while reading through the documentation.

motwok committed:
add yaml to test script

motwok commented:
I had the same ideas.

After a long time of brainstorming i came to the conclusion that we had to people to show the posibility first.
So i set the main target to on the pur data transfer with the automation actions.

Second targets are security, large packets, and auto channeling.

If this is done we can talk about creating something that looks like a component and uses the current actions and events to mimic this.

I did not mention this because i am completely unaware how to attach the actions to an event, i have figure it out first.

Emmo

motwok committed:
fix yaml of device 4

motwok committed:
Merge branch 'dev' into wifi_now_component

motwok committed:
Merge branch 'wifi_now_component' of https://github.com/motwok/esphome into wifi_now_component

motwok committed:
Add WIFI_BASIC_COOP define

motwok committed:
setting Setup Priority higher than WIFI
overworking setup add better messages and remove ESP_ERROR_CHECK
Fix some comparison bugs

motwok committed:
fix lint message

motwok committed:
fix some lint messages
add set_warning when send fails.

motwok committed:
add codeowners

motwok committed:
fix typo and add some template files

motwok committed:
fix lint message

motwok committed:
codeowners fixed

motwok committed:
removing empty cpp files

motwok committed:
Merge branch 'dev' into wifi_now_component

penright commented:
I saw where you mention this on the dev channel. It has been some time since I messed with the ESP now somewhat. If I remember right there was some issue with matching the channel with the same one that was used for the AP WIFI. 
Did you have that issue in your implementation? 
I assume this is the code. https://github.com/motwok/esphome/tree/wifi_now_component/esphome/components/wifi_now

OttoWinter commented on esphome/components/wifi_now/md5sum.py:
```diff
@@ -0,0 +1,21 @@
+
+def get_md5sum(text):
+    import subprocess
+    process = subprocess.Popen(
```
No need to calculate md5 with an external process.

Please see [hashlib](https://docs.python.org/3/library/hashlib.html) - `h = hashlib.new('md5'); h.update(text); h.hexdigest()`

motwok replied:
Hello Otto,

thanks for the review.
I will change the code to hashlib

motwok replied:
Done with commit 93ac344a8e91dfa193836b05b0707fc4c1ba368e

OttoWinter commented on script/test:
```diff
@@ -10,3 +10,8 @@ esphome tests/test1.yaml compile
 esphome tests/test3.yaml compile
 esphome tests/test4.yaml compile
+
+esphome tests/esp_now_device1.yaml compile
```
Please only add 1 test file max (or if possible even put it in the other files).

Every new YAML file here increases our continuous integration time, so we try to keep this as low as possible.

motwok replied:
Hello Otto,
I understand the impact on ci.

Curently 4 Scenarios here

esp8266 w/ wifi
esp8266 w/o wifi
esp32 w/ wifi
esp32 w/o wifi

I will reduce to 2 Scenarios, esp32 and esp8266 use different c++ code this is the meaningfull minimum.

Emmo

motwok replied:
Done with commit cee43f8e4cd15e44a559d3c70c78069a164ed148 and previous

motwok committed:
Merge branch 'dev' into wifi_now_component

motwok committed:
fixing issue 1 (from motwok/mo2000_esphome_issues)

motwok committed:
Changed get_md5_sum to use hashlib

as intended by Otto Winter on March 20, 2021
see https://github.com/esphome/esphome/pull/1290#discussion_r598152768

motwok commented:
Sorry i am anwering late... priorities, job first, 

> I saw where you mention this on the dev channel. It has been some time since I messed with the ESP now somewhat. If I remember right there was some issue with matching the channel with the same one that was used for the AP WIFI.
> Did you have that issue in your implementation?
> I assume this is the code. https://github.com/motwok/esphome/tree/wifi_now_component/esphome/components/wifi_now

yes there is an issue with the channel, currently i work arround using a fixed channel on 2.4GHz Wifi. 

I had a kind of autochanneling in mind that fixes the problem as long there is only on AP.
Having two AP on different channels is a challenge, but i dont see a use case for this (and many simple  workarounds).

Emmo

motwok committed:
Lint demands those BS changes

motwok committed:
Removed NonWifi Tests

as intended by Otto Winter on March 20, 2021
see https://github.com/esphome/esphome/pull/1290#discussion_r598152967

motwok committed:
fix Testnames

Primo83 commented:
Hey @motwok, will the work be continued?

motwok commented:
> Hey @motwok, will the work be continued?

sure, when my job gives me time.

motwok committed:
Merge branch 'dev' into wifi_now_component

motwok commented:
Push for first annual

balk77 commented:
Hi, I am interested in testing this PR. Never done that though. How do I do that? Is this in a beta docker image or do I need to install from scratch? Any special testing requests?

mjkl-gh commented:
> Hi, I am interested in testing this PR. Never done that though. How do I do that? Is this in a beta docker image or do I need to install from scratch? Any special testing requests?

https://esphome.io/components/external_components.html?highlight=pr 

under "# equivalent shorthand for GitHub pull request" it shows you how to use a PR for a specific device

Bitbanger1900 commented:
Hi @motwok

I am hoping for a long time to see Wifi-Now integrated into Esphome and i want to say thanks for your work!

i managed to get your code compiled but somehow i can't get it to work with a Temperature sensor. Is it possible to get
an example of a sending and receiving yml with a Temperature sensor? 

Regards,
Bitbanger.

balk77 commented:
I managed pull in the code by adding this to my configuration. Running the DEV Home Assistant addon.
```
external_components:
  - source: github://pr#1290
    components: [ wifi_now ]
esphome:
  name: ${node_name}
  platform: ESP32
  board: nodemcu-32s
```
The compilation fails, also after cleaning the build files. @Bitbanger1900 what kind of board are you using? Mine is ESP32.
```
Compiling /data/aanrechtdimmerknop/.pioenvs/aanrechtdimmerknop/src/esphome/core/scheduler.cpp.o
In file included from src/esphome/components/wifi_now/receive_trigger.h:7:0,
                 from src/esphome/components/wifi_now/receive_trigger.cpp:1:
src/esphome/core/base_automation.h: In member function 'bool esphome::ForCondition<Ts>::check_internal()':
src/esphome/core/base_automation.h:69:37: error: there are no arguments to 'millis' that depend on a template parameter, so a declaration of 'millis' must be available [-fpermissive]
       this->last_inactive_ = millis();
                                     ^
src/esphome/core/base_automation.h:69:37: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
src/esphome/core/base_automation.h: In member function 'bool esphome::ForCondition<Ts>::check(Ts ...)':
src/esphome/core/base_automation.h:76:19: error: there are no arguments to 'millis' that depend on a template parameter, so a declaration of 'millis' must be available [-fpermissive]
     return millis() - this->last_inactive_ >= this->time_.value(x...);
                   ^
*** [/data/aanrechtdimmerknop/.pioenvs/aanrechtdimmerknop/src/esphome/components/wifi_now/receive_trigger.cpp.o] Error 1
========================== [FAILED] Took 6.04 seconds ==========================
```
I am not a programmer, don't know how to fix this.

Bitbanger1900 commented:
@balk77 

uit mijn hoofd.. (i was on holiday..)
- git clone https://github.com/motwok/esphome
- zip esphome.zip esphome -r
- pip3 install esphome.zip. 

after that: esphome version = Version: 1.18.0-dev

I have esp32dev as board but with nodemcu-32s i also have no error's.

motwok committed:
Merge branch 'esphome:dev' into wifi_now_component

manutenfruits commented:
I think enabling this feature would truly mean opening esphome to a myriad of new DIY devices that can be battery powered. It has big demand and I really wish some kind sould would look into it. If the original author doesn't have bandwidth, Nabu Casa should look into staffing the project.

motwok commented:
Currtently this PR is outdated!

I am working on a complete new version...
